Just skip 'already exists' in CREATE TABLE IF NOT EXISTS PARTITION OF

naisila opened the pull request:
DESCRIPTION: Just skip 'already exists' in CREATE TABLE IF NOT EXISTS PARTITION OF, don't raise "already distributed" error

We enter `create_distributed_table` part if:
* current parent in the query is a Citus table
* relation is not already distributed (not a Citus table)
* relation is a partition
* relation's parent is the current parent

The last three points are important in an `IF NOT EXISTS` statement.

Fixes #2087

halilozanakgul commented on src/backend/distributed/commands/table.c:
```diff
@@ -224,9 +224,12 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
-		CreateDistributedTable(relationId, parentDistributionColumn,
-							   parentDistributionMethod, ShardCount,
-							   parentRelationName, viaDeprecatedAPI);
+		if (!(IsCitusTable(relationId) && (createStatement->if_not_exists)))
```
This check doesn't look right to me.

It means we will distribute the table if it is not a Citus table OR there is no `IF NOT EXISTS` statement. What if there is an `IF NOT EXISTS` statement and the table exists but it is just a regular Postgres table (that is not a partition). Won't this code just try to distribute it?

naisila replied:
Before there was no check in the code, I put one specific for the issue.
I guess before that check, we would enter that part of the code in this condition:
```
/*
  * If a partition is being created and if its parent is a distributed
  * table, we will distribute this table as well.
  */
```
So something that is not a partition shouldn't enter here.

If the partition already exists but is not distributed: actually this should never be the case for a parent distributed table. As I have seen from the tests, the codebase ensures that all partitions of a distributed table are distributed as well.

agedemenli replied:
* We can move lines [222, 225] to scope of this if statement, since we don't need/use them unless this if condition is satisfied.
* This if condition is not very readable. Adding comments would be nice.
* For Ozan's crashing case, `PreprocessCreateTableStmtPartitionOf` skips the creation since `if_not_exists` is true and a table with this name is already created. However, for this function, we don't know whether it's just created in `PreprocessCreateTableStmtPartitionOf` or was already existing. So I haven't tested it but maybe not calling `PreprocessCreateTableStmtPartitionOf` in such cases might be a solution

naisila replied:
By the way, couldn't find any `PreprocessCreateTableStmtPartitionOf`

agedemenli replied:
Sorry, didn't check that. But still, the case is the same. We execute `CREATE TABLE` queries (and others) **before** the Postprocess function.

halilozanakgul reviewed (changes requested):
```
CREATE TABLE prt (a INT) PARTITION BY RANGE (a);
SELECT create_distributed_table ('prt', 'a');

CREATE TABLE prt_1 (a INT);

CREATE TABLE IF NOT EXISTS prt_1 PARTITION OF prt FOR VALUES FROM (1) TO (5);
```

This crashes!

agedemenli commented on src/backend/distributed/commands/table.c:
```diff
@@ -219,13 +219,17 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
+		 * If the relation already exists and it's an IF NOT EXISTS stmt,
+		 * just skip like in plain PG.
+		 */
+		if (!(RelationExists(relationId) && (createStatement->if_not_exists)))
```
This causes not distributing new created tables. I mean, for a valid create table command, it gets created after Preprocess function and before Postprocess function. It doesn't matter it's an already existing table or not, `RelationExists(relationId)` will always return true. And when `if_not_exists` is also true, the if condition will not be satisfied, hence the table in a valid creation command will not get distributed.
Example:
```
CREATE TABLE "test2" (
   "cdc" INTEGER,
   "product" INTEGER)
PARTITION BY LIST ("cdc");
SELECT create_distributed_table('test2', 'product');
CREATE TABLE IF NOT EXISTS "test3" PARTITION OF "test2" FOR VALUES IN (1); 
```
Then you can see by connecting to a worker with `psql -p 9701` that it's not distributed

naisila replied:
Makes sense, I forgot about that 👍

agedemenli replied:
The problem here is, when we run `CREATE TABLE IF NOT EXISTS "test3" PARTITION OF "test2" FOR VALUES IN (1);` either we already have a table named test3, or we don't; there will be a table named `test3` when it comes to your if statement. So we couldn't know it's an already existing table or the one we just created (Between Preprocess and Postprocess functions). This is the reason for `(RelationExists(relationId)` is pointless. It will always return true.

naisila replied:
Yes, I understood what you meant in the first time, I just forgot about that :)

> Then you can see by connecting to a worker with psql -p 9701 that it's not distributed

I was saying for this one, I can check directly from coordinator in pg_dist_partition - test3 will not be there in your example case.

agedemenli commented:
We might need a deparser and creating DDLJobs for this task. Because this way we'll never know if the table with the given name was already existing, or just created with the current command. @halilozanakgul what do you think?

naisila commented:
I think I can make use of `isPartitionTable` but let me try

naisila commented:
Update:
* If the table with the given name was just created, that means it was a partition so I check that with `PartitionTable()` function.
* I removed the `if_not_exists` condition since if naming errors were not prevented by PG before it means that there was an `IF NOT EXISTS` part.
* Also, I tried checking before `Postprocess...` rather than inside `Postprocess...`, which one seems more readable?

I also added tests, they all pass but let me know if you have something in mind that might crash.

agedemenli commented on src/backend/distributed/commands/table.c:
```diff
@@ -186,8 +187,15 @@ PostprocessCreateTableStmt(CreateStmt *createStatement, const char *queryString)
+		 */
+		if (!(IsCitusTable(relationId)) && PartitionTable(relationId))
+		{
+			PostprocessCreateTableStmtPartitionOf(createStatement, queryString);
```
It still tries to distribute an existing table with the same name, which is a partition of some other table which is not distibuted. Because the existing table with the same name, satisfies these two conditions; it's not a citus table (not distributed) and it's a partition table. But errors out when distributing, because its parent relation is not distributed.

Example:
```
CREATE TABLE "test1" (
   "cdc" INTEGER,
   "product" INTEGER)
PARTITION BY LIST ("cdc");

CREATE TABLE "test2" (
   "cdc" INTEGER,
   "product" INTEGER)
PARTITION BY LIST ("cdc");

SELECT create_distributed_table('test1', 'product');

CREATE TABLE IF NOT EXISTS "test3" PARTITION OF "test2" FOR VALUES IN (1);

CREATE TABLE IF NOT EXISTS "test3" PARTITION OF "test1" FOR VALUES IN (1);

NOTICE:  relation "test3" already exists, skipping
ERROR:  cannot distribute relation "test3" which is partition of "test2"
DETAIL:  Citus does not support distributing partitions if their parent is not distributed table.
HINT:  Distribute the partitioned table "test2" instead.
```

naisila replied:
Makes sense, I will add a check for that

naisila commented on src/test/regress/sql/multi_partitioning.sql:
```diff
@@ -1155,13 +1155,25 @@ ALTER TABLE partitioning_test ATTACH PARTITION partitioning_test_2011
 SELECT parent_table, partition_column, partition, from_value, to_value FROM time_partitions;
 
+-- create the same partition to verify it behaves like in plain PG
+CREATE TABLE partitioning_test_2011 PARTITION OF partitioning_test FOR VALUES FROM ('2011-01-01') TO ('2012-01-01');
+CREATE TABLE IF NOT EXISTS partitioning_test_2011 PARTITION OF partitioning_test FOR VALUES FROM ('2011-01-01') TO ('2012-01-01');
+CREATE TABLE IF NOT EXISTS partitioning_test_2012 PARTITION OF partitioning_test FOR VALUES FROM ('2012-01-01') TO ('2013-01-01');
+
+-- create the same table but that is not a partition and verify it behaves like in plain PG
+CREATE TABLE not_partition(time date);
+CREATE TABLE not_partition PARTITION OF partitioning_test FOR VALUES FROM ('2011-01-01') TO ('2012-01-01');
+CREATE TABLE IF NOT EXISTS not_partition PARTITION OF partitioning_test FOR VALUES FROM ('2011-01-01') TO ('2012-01-01');
+DROP TABLE not_partition;
```
You mean for `partitioning_test_2012` right? I will add a check in `pg_dist_partition`.

agedemenli commented on src/test/regress/expected/multi_partitioning.out:
```diff
@@ -1958,12 +1958,33 @@ SELECT parent_table, partition_column, partition, from_value, to_value FROM time
+NOTICE:  relation "partitioning_test_2011" already exists, skipping
+-- verify we can create a partition that doesn't already exist with IF NOT EXISTS
+CREATE TABLE IF NOT EXISTS partitioning_test_2012 PARTITION OF partitioning_test FOR VALUES FROM ('2012-01-01') TO ('2013-01-01');
+SELECT COUNT(*) FROM pg_dist_partition WHERE logicalrelid = 'partitioning_test_2012'::regclass;
+ count
+-------
+     1
+(1 row)
```
Instead, we can verify this by checking the worker. You can connect to worker node and list shards, in test sql file by using `\c - - - :worker_1_port`. See `alter_index.sql` for an example.

naisila replied:
I am familiar with testing in the worker nodes from previous issues, but in this particular case I think using `pg_dist_partition` is more consistent here, considering the rest of `multi_partitioning.sql`.
For more consistence, I will modify my use of `pg_dist_partition` and make it the same as the other uses in the same file.

agedemenli commented on src/backend/distributed/commands/table.c:
```diff
@@ -219,14 +219,22 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
+		 * of the current parent, just skip like in plain PG.
+		 */
+		if (!(IsCitusTable(relationId)) && PartitionTable(relationId) &&
+			(PartitionParentOid(relationId) == parentRelationId))
```
minor: you could explain this `PartitionParentOid(relationId) == parentRelationId` check in the comment, since it's a little bit difficult to see when reading this code piece.

naisila replied:
It is already explained in the comment
`...or if the table is not a partition of the current parent...`

agedemenli reviewed (commented):
Looks good. Thank you!

halilozanakgul commented on src/backend/distributed/commands/table.c:
```diff
@@ -221,14 +221,22 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
+		 * If that partition is already distributed, or if the table is not a partition
+		 * of the current parent, just skip like in plain PG.
+		 */
+		if (!(IsCitusTable(relationId)) && PartitionTable(relationId) &&
```
Can you remove the brackets around the `IsCitusTable`. They made me think `!` will cover more checks.

halilozanakgul commented on src/backend/distributed/commands/table.c:
```diff
@@ -221,14 +221,22 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
 	if (IsCitusTable(parentRelationId))
```
I think we could combine the new if with this one so it will be clearer that we distribute when:

- parentRelation is a Citus table
- relation is not a Citus table
- relation is a partition
- relation's parent is parentTable

naisila replied:
So you suggest a single `if` statement with 4 parts?

halilozanakgul replied:
Yes 👍

halilozanakgul commented on src/test/regress/sql/multi_partitioning.sql:
```diff
@@ -1155,13 +1155,35 @@ ALTER TABLE partitioning_test ATTACH PARTITION partitioning_test_2011
+-- verify it skips when the partition with the same name belongs to another table
+CREATE TABLE another_table(id int, time date) PARTITION BY RANGE (time);
+CREATE TABLE partition_of_other_table PARTITION OF another_table FOR VALUES FROM ('2014-01-01') TO ('2015-01-01');
+CREATE TABLE IF NOT EXISTS partition_of_other_table PARTITION OF partitioning_test FOR VALUES FROM ('2014-01-01') TO ('2015-01-01');
```
For completeness can you add the same test without the `IF NOT EXISTS`

onderkalaci commented on src/backend/distributed/commands/table.c:
```diff
@@ -214,13 +214,19 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
+	 * and that partition's parent is the current parent from parentRelationId
 	 */
-	if (IsCitusTable(parentRelationId))
+	if (IsCitusTable(parentRelationId) && !IsCitusTable(relationId) &&
```
I think in terms of correctness, this is a pretty comprehensive check, preventing several edge cases, which is great. (Also, great test coverage!)

In terms of the code, I see some potential improvement areas, please feel free to consider the following.

I think with this check, we are trying to do multiple unrelated checks at once and also behaving `if_not_exists`  very implicitly. 

Instead, it'd be easier to follow if we make these checks explicitly. Please see the diff: https://github.com/citusdata/citus/compare/issue2087...onder_issue2087?expand=1

I think that you don't have to (and probably shouldn't as the comments have to be be improved) the proposed diff above, but something along with these lines can be nicer in terms of code readability.

naisila replied:
It seems more readable to completely separate the IF NOT EXISTS case. Will go for it :+1:
